### ISSUE TITLE
Spurious errors for lack of specialization involving pre 3.12 typealias use

### ISSUE THREAD

**Issue by mikeshardmind** [microsoft/pyright#10679]
Code sample in [pyright playground](https://pyright-play.net/?strict=true&code=JYWwDg9gTgLgBAYwgG2QUwTYEB2BnAOgEMAjBAWAChRJY48BPPKgMyghDhgbGBwHM4NaPAAqPNAEFkwIngA0cAMpoAjgFc0OBGkUB5MFlxFkVKsBb0mBAG5ooebDgD6fFhDgA%2BALxwAFADMigCMAEwAlABcVHCxcABC6iws9pFw4mBSMnJwvkioGEb4xGQEicn2VGjIeGjRlHEJSSlQaRlZsni5cCQMMGjMlGaUALJEYLwCbRLSnd356JhOhKQIBGMTfPzDSgCKADLA-QBqJprdAHK4aHAAPkI48PcsyBBET00VUHf0MFDDVAQyDkXQAXvYICRXiR6o0AEQIgCSlgYEHUcAA7kRHlwPHxanQiD1oQ8YB4iewMYpUeicGg0AATXExOJ4dQTZAMLgACxuWiwUBuUMhDzg1zg-AgBFiAFVHAIecAukCQWL-g04gADFjqbRFOAAHm5MBgYDwkQA9BaMTaCHhVDJ%2BgRoPwLcCBM4kIKddoCMaQMgAMTg9jCkieTXOZws2ImV4Id4DHk3PBgIg6UUwXlwBnvUhyG4sYCoLZCeAYo7cuAhiADAjDRoATTRiGxyZwYrsUAxUCONzTsC6fHoIDj9kQ3N1AGsFJjK2j4NyiDYtjHcTm8fA43B3mLtGhpels5FvZFNWgAB5EcDoQ2X69gdDOMOuCAR%2BjciAYwaNT8Y9fqLUZYEAicINnEDJoJYUZ8EcUZ%2BLUyAsIojjgmkfAwFEa7wgikQDte9DAGhcAXOoIAkOOECWDWjI9H0SZkoggqJqB2FxAQHHgbEkGWOgAhZgh1QsOEcAALSeKSsKNHEoFKERNxUdWELEiKw69P0hCsRq0lwBx9ZDJQ8R8AyWxdL4KgaFoOgANp7IcJxnDc9w1mGAC6PwbJM-DWXgfyKHZfanMg5zORCbnDCqeBdAAwuoDjQFJsR6QClCRTFuB0ksuD1GuPFwJeGDqP0glIShMCJiA-Lmr8UCKCQxmmWkBhFCY1lGTgJkCHg7m%2BFcdKKAAVIoCYuAmCC8mkJAQCg3SiFAmiKGAgoDmgzgvEQ-DVRh3QAAyKJej5EHw6E4r4onBCJ4lwLF8WtGxsSgQAohehUacmvwVVVcCAaWWYpuyj7ALR9UdaZh7Pa9Tj3XAgowHF%2BCYryHZ-XARYOPAlJCF0y5HcCUIKd826%2BZ9jxdEudjQ0gN5oP0DKHn4kiIHFeDQFjO5FRw7zAAmqBcpC5V8Iy4T6dDABirNLRto5dDSH5osgTKATckSVVmZ6S-wo6aoj3NVkuHXoF0GgQFgCrYgylPptypZeoKmCciL2nSSoyuq9yZ43SzUAEAVCBFWg2vuN8IDQDckECzUijm%2B9J5QWe97U3eV7U84vtFVszj2sgEYgQi0N6UAA)

```python
import collections.abc
import sys
from typing import TypeAlias, Sequence, Optional

if sys.version_info >= (3, 12):
    Buffer: TypeAlias = collections.abc.Buffer
else:
    Buffer: TypeAlias = bytes


Mapping: TypeAlias = collections.abc.Mapping

SQLiteValue = None | int | float | Buffer | str


class zeroblob:
    """If you want to insert a blob into a row, you need to
    supply the entire blob in one go.  Using this class or
    `function <https://www.sqlite.org/lang_corefunc.html#zeroblob>`__
    allocates the space in the database filling it with zeroes.

    You can then overwrite parts in smaller chunks, without having
    to do it all at once.  The :ref:`example <example_blob_io>` shows
    how to use it."""

    def __init__(self, size: int):
        """:param size: Number of zeroed bytes to create"""
        ...

    def length(self) -> int:
        """Size of zero blob in bytes."""
        ...

Bindings = Sequence[SQLiteValue | zeroblob] | Mapping[str, SQLiteValue | zeroblob]

class Cursor:
    ...


class Connection:

    def execute(self, statements: str, bindings: Optional[Bindings] = None, *, can_cache: bool = True, prepare_flags: int = 0, explain: int = -1) -> Cursor:
        ...
```

```
Expected type arguments for generic class "Mapping"  (reportMissingTypeArgument)
Type "Mapping[Unknown, Unknown]" is already specialized  (reportInvalidTypeArguments)
Type of parameter "bindings" is partially unknown
  Parameter type is "Sequence[int | float | Buffer | str | zeroblob | None] | Mapping[Unknown, Unknown] | None"  (reportUnknownParameterType)
```

Above example is a minimized form of what occurs with use of [apsw](https://github.com/rogerbinns/apsw)

Error occurs in pyright 1.1.402 (As well as current playground), this appears to be a regression, though I'm not sure what version this began erroring in.

**Comment by erictraut** [comment]
Pyright is working correctly here, so I don't consider this a bug.

You're defining a type alias named `Mapping` that is not generic — and therefore does not accept any type arguments. The declaration is equivalent to the following because the type parameters for `collections.abc.Mapping` don't have default values, so they default to `Any` if left unspecified.

```python
Mapping: TypeAlias = collections.abc.Mapping[Any, Any]
```

If your intent is to define a generic type alias, you would need to do this:

```python
V_co = TypeVar("V_co", covariant=True)
K = TypeVar("K")

Mapping: TypeAlias = collections.abc.Mapping[K, V_co]
```

Or if you are using Python 3.12, you can use the new syntax:

```python
type Mapping[K, V] = collections.abc.Mapping[K, V]
```

**Comment by mikeshardmind** [comment]
pyright previously accepted this, and mypy still does. 

I'm not sure I follow the explanation here, as if that's how pyright is interpreting it, the first error seems to contradict that as it is states it is expecting to be parameterized.

The Typevar use without any corresponding use also doesn't seem correct, in fact, pyright rejects the same kind of use that ends up not being solved in `Callable`, requiring Protocol use for functions returning generic functions.

**Comment by mikeshardmind** [comment]
The specification also says nothing about this, only saying:

> Type aliases may be defined by simple variable assignments:

> Or by using typing.TypeAlias:

Implying the two to have the same behavior, and not providing any special rules for typevariables in old type aliases. The code above works when the `TypeAlias` hint is removed.

**Comment by erictraut** [comment]
> the first error seems to contradict that

By "first error", I presume you're talking about the `reportMissingTypeArgument` error? That's because `collections.abc.Mapping` _is_ a generic class that expects two type arguments, and you haven't supplied any type arguments.

> pyright previously accepted this

By "this", I presume you're referring to the `reportInvalidTypeArguments` error. Yes, there were bugs in older versions of pyright where it failed to generate an error in this case. I don't recall which version fixed this, but it was a long time ago.

> The specification also says nothing about this ...

The typing spec is clear about how type expressions should be evaluated. The type expression `collections.abc.Mapping` (with no subscript) is interpreted as the type `collections.abc.Mapping[Any, Any]`. This type cannot be further specialized because it is already specialized.

PEP 613 introduced `TypeAlias` to unambiguously differentiate between type alias definitions and global variable assignment statements. Prior to PEP 613, the difference between the two was murky and under-specified. Type checkers needed to apply various heuristics to guess what the user meant. With the introduction of `TypeAlias`, this ambiguity was eliminated. The RHS of a `TypeAlias` definition is always evaluated as a type expression. All type expression rules apply in this case — including the replacement of unspecified type arguments with `Any` (or with the default type argument value if a default has been specified).

**Comment by mikeshardmind** [comment]
>  The type expression collections.abc.Mapping (with no subscript) is interpreted as the type collections.abc.Mapping[Any, Any]

Okay, accepting that, the first error then is spurious, as it doesn't require parameterization, and has a meaning then? Any interpretation here should involve at least one of these errors being spurious.

**Comment by erictraut** [comment]
By "first error", are you referring to the `reportMissingTypeArgument` diagnostic? That's not a spurious error. The `reportMissingTypeArgument ` check is working as designed here.

The `reportMissingTypeArgument` is a strict-mode check that is not enabled by default. I presume that you chose to enable it in your project. If you don't want to see diagnostics related to this check, you can leave it disabled.

**Comment by mikeshardmind** [comment]
It wasn't clear to me that it was working as intended. The documentation for this says:

> reportMissingTypeArgument [boolean or string, optional]: Generate or suppress diagnostics when a generic class is used without providing explicit or implicit type arguments. The default value for this setting is "none".

and you've indicated that pyright is interpreting this case as implicitly having type arguments of [Any, Any].

Not sure what the purpose of this rule is given that the specification currently *always* implicitly treats them as Any (well, Any as the default default now that defaults are explictly supported), so maybe it's a documentation issue? It doesn't seem like a rule based on the specification at any rate, and therefore something I'm surprised to see an error for.
